ExpressLRS/ExpressLRS
# Airport

### What is Airport?
Airport provides a new firmware option that allows you to turn a regular ExpressLRS TX and RX pair into a bi-directional transparent serial link, over the air.
The TX module is intended to connect via USB to a laptop, and the RX connects to a free UART on your FC, as per usual. This then allows serial comms between the peers, in any protocol you wish to use, for example: MAVLINK (Ardupilot), MSP (Betaflight and INAV), or any other telemetry protocol you might use.

### What hardware do I need to try this?
The Airport option completely replaces the RC link, and repurposes it as a data link. This means you need to run 2x TXs and 2x RXs on the ground and air, respectively. One TX+RX pair sends your normal RC link data, just as it does it does now, and the other TX+RX pair sends the serial data.
![airport-diagram](https://user-images.githubusercontent.com/59873510/198181617-85335145-1972-4d4b-acb3-c3b744a5ab61.png)

The feature should compile for any supported ELRS target, however the TX module makes sense to be an external one, so you can connect it via USB to a laptop, and it is helpful to have an OLED screen on the TX module so you can change packet rates / power output without plugging it into a handset JR bay. It should compile for 868/900/2.4 targets, on both STM and ESP architectures, so all RXs with a full UART (both tx and rx pads) should work.

### What baud rate should I use?
The baud rate you select to use should be the same on the TX and RX, and it should also be slow enough such that the packet rate can send the data.
For example, at a packet rate of 500hz LoRa, and a 1:2 hard-coded TLM rate, we can send 500hz / 2 * 5bytes = 1250bytes/s.
A baud rate of 19200 allows for a data rate of 19200 * 0.8 / 8bits/byte = 1920bytes/s. This is too much data for 500hz, since 1920 > 1250.
A baud rate of 9600 allows for a data rate of 9600* 0.8 / 8bits/byte = 960bytes/s. This is an acceptable rate since 960 < 1250.

### How do I test this PR?
Long term, this will be an option you can enable in the ELRS Configurator.
For now though, the user define should be used via the manual FW build process (read: dev friendly only ATM).
Enable the `-DUSE_AIRPORT_AT_BAUD=9600` option by removing the # at the beginning of the line in `user_defines.txt`, and set your binding phrase user define to something different to what you use for normal ELRS RC control.
Build and flash a TX and RX pair using the above options, and connect RX to a spare UART, and TX to a PC via USB.
Configure your FC to talk via whatever protocol you prefer, for example if using Ardupilot, configure the UART for MAVLINK protocol.
Open the connection from the Configurator / GCS system on the PC.
Profit!
